src/components/lab/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,320 @@
+        onBlur,
+        onKeyDown,
+        onKeyUp,
+        allowMouseWheel = false,
+        allowDecimal = false,
+        className,
+    } = {...defaultBaseTextInputProps, ...props};
+
+    const {
+        min,
+        max,
+        step: baseStep,
+        value,
+        defaultValue,
+        shiftMultiplier,
+    } = getInternalVariables({
+        min: externalMin,
+        max: externalMax,
+        step: externalStep,
+        shiftMultiplier: externalShiftMultiplier,
+        allowDecimal,
+        value: externalValue,
+        defaultValue: externalDefaultValue,
+    });
+
+    const [inputValue, setInputValue] = React.useState(
+        getStringValue(value) ?? getStringValue(defaultValue),
+    );
+
+    React.useEffect(() => {
+        const stringPropsValue = getStringValue(value);
+        setInputValue((currentInputValue) => {
+            if (!areStringRepresentationOfNumbersEqual(currentInputValue, stringPropsValue)) {
+                return stringPropsValue;
+            }
+            return currentInputValue;
+        });
+    }, [value]);
+
+    const clamp = true;
+    const [step, setStep] = React.useState(baseStep);
+    const [active, setActive] = React.useState(false);
+
+    const state = getInputControlState(validationState);
+
+    const canIncrementNumber =
+        clampToNearestStepValue({value: value + step, step, min, max}) > value;
+
+    const canDecrementNumber =
+        clampToNearestStepValue({value: value - step, step, min, max}) < value;
+
+    const innerControlRef = React.useRef<HTMLInputElement>(null);
+    const fieldRef = useFormResetHandler({
+        initialValue: defaultValue,
+        onReset: onUpdate ?? voidFunction,
+    });
+    const handleRef = useForkRef(props.controlRef, innerControlRef, fieldRef);
+
+    const handleIncrement = () => {
+        if (canIncrementNumber) {
+            onUpdate?.(clampToNearestStepValue({value: value + step, step: baseStep, min, max}));
+        }
+    };
+
+    const handleDecrement = () => {
+        if (canDecrementNumber) {
+            onUpdate?.(clampToNearestStepValue({value: value - step, step: baseStep, min, max}));
+        }
+    };
+
+    const handleWheel: React.WheelEventHandler<HTMLInputElement> = (e) => {
+        const delta = e.shiftKey ? e.deltaX : e.deltaY;
+        if (delta > 0) {
+            handleIncrement();
+        } else if (delta < 0) {
+            handleDecrement();
+        }
+    };
+
+    const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+        if (e.key === KeyCode.SHIFT) {
+            setStep(baseStep * shiftMultiplier);
+        } else if (e.key === KeyCode.ARROW_DOWN) {
+            handleDecrement();
+        } else if (e.key === KeyCode.ARROW_UP) {
+            handleIncrement();
+        }
+        onKeyDown?.(e);
+    };
+
+    const handleKeyUp: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+        if (e.key === KeyCode.SHIFT) {
+            setStep(baseStep);
```

Comment:
We also should reset step on input's blur, otherwise it stucks on Tab/Shift+Tab navigation